File: eclipse-initial/src/native/org.eclipse.ice.reactor.native/pwr/IncoreInstrument.cpp
/*******************************************************************************
* Copyright (c) 2012, 2014 UT-Battelle, LLC.
* All rights reserved. This program and the accompanying materials
* are made available under the terms of the Eclipse Public License v1.0
* which accompanies this distribution, and is available at
* http://www.eclipse.org/legal/epl-v10.html
*
* Contributors:
*   Initial API and implementation and/or initial documentation - Jay Jay Billings,
*   Jordan H. Deyton, Dasha Gorin, Alexander J. McCaskey, Taylor Patterson,
*   Claire Saunders, Matthew Wang, Anna Wojtowicz
*******************************************************************************/

#include "IncoreInstrument.h"
#include "../HDF5LWRTagType.h"
#include <string>
#include <vector>
#include <HdfReaderFactory.h>
#include <HdfWriterFactory.h>
#include <H5Cpp.h>
#include "../UtilityOperations.h"
#include "../LWRComponent.h"
#include "../Ring.h"
#include <ICEObject/Identifiable.h>

using namespace ICE_Reactor;

/**
 * The Copy Constructor
 */
IncoreInstrument::IncoreInstrument(IncoreInstrument & arg) : LWRComponent(arg) {

    //begin-user-code

    //Clone for getting rings and tubes!
    std::shared_ptr<Identifiable> identRing = arg.thimble.get()->clone();

    //Dynamic cast up to a ring (which will preserve the clone operation on stack without calling new!
    std::shared_ptr<Ring> ring = std::dynamic_pointer_cast<Ring> (identRing);
    this->thimble = ring;

    //end-user-code

}

/**
 * The Destructor
 */
IncoreInstrument::~IncoreInstrument() {
    //TODO Auto-generated method stub
}

/**
 * The nullary Constructor.
 */
IncoreInstrument::IncoreInstrument() {

    //begin-user-code

    //Setup defaults
    this->name = "Instrument 1";
    this->description = "Default Instrument";
    this->id = 1;
    this->HDF5LWRTag = INCORE_INSTRUMENT;
    this->thimble = std::shared_ptr<Ring> (new Ring("thimble"));

    //end-user-code
}

/**
 * A parameterized Constructor.
 *
 * @param the name
 * @param the thimble
 */
IncoreInstrument::IncoreInstrument(const std::string name, std::shared_ptr<Ring> thimble) {
    //begin-user-code

    //Setup defaults
    this->name = "Instrument 1";
    this->description = "Default Instrument";
    this->id = 1;
    this->HDF5LWRTag = INCORE_INSTRUMENT;
    this->thimble = std::shared_ptr<Ring> (new Ring("thimble"));

    //Call setters
    IncoreInstrument::setName(name);
    IncoreInstrument::setThimble(thimble);

    //end-user-code
}

/**
 * Sets an empty thimble tube used a boundary between the detector and the reactor.
 *
 * @param the thimble to set
 */
void IncoreInstrument::setThimble(std::shared_ptr<Ring> thimble) {

    // begin-user-code

    //if thimble is not Null set value
    if (thimble.get() != NULL) {

        this->thimble = thimble;
    }

    // end-user-code

}

/**
 * Returns an empty thimble tube used a boundary between the detector and the reactor.
 *
 * @return the thimble
 */
std::shared_ptr<Ring> IncoreInstrument::getThimble() {

    // begin-user-code

    return this->thimble;

    // end-user-code
}

/**
 * This operation writes HDF5 Attributes to the metadata of h5Group in the h5File. If the h5Group is null or h5File is null or can not be opened, then false is returned. If the operation fails to write all Attributes, then false is returned. Otherwise, true is returned.
 *
 * @param The File
 * @param The Group
 *
 * @return True if successful, false otherwise
 */
std::vector< std::shared_ptr<ICE_IO::IHdfWriteable> >  IncoreInstrument::getWriteableChildren() {
    //begin-user-code

    //Get the children in super
    std::vector< std::shared_ptr<ICE_IO::IHdfWriteable> > children = LWRComponent::getWriteableChildren();

    //Add thimble to list
    children.push_back(this->thimble);

    //Return list
    return children;

    //end-user-code

}

/**
 * Overrides the equals operation to check the attributes on this object with another object of the same type.  Returns true if the objects are equal.  False otherwise.
 *
 * @param object to compare to
 *
 * @return true if equal, false otherwise
 */
bool IncoreInstrument::operator==(const IncoreInstrument& otherObject) const {

    // begin-user-code

    return LWRComponent::operator ==(otherObject) && this->thimble.get()->operator ==(*otherObject.thimble.get());

    // end-user-code
}

/**
 * Deep copies and returns a newly instantiated object.
 *
 * @return the newly instantiated object
 */
std::shared_ptr<ICE_DS::Identifiable> IncoreInstrument::clone() {

    // begin-user-code

    // Local Declarations
    std::shared_ptr<IncoreInstrument> component(new IncoreInstrument (*this));

    // Return the newly instantiated object
    return component;

    // end-user-code

}

/**
 * This operation reads IHdfReadable child objects. If this IHdfReadable has no IHdfReadable child objects, false is returned.
 *
 * @param the readable child
 *
 * @return true if successful, false otherwise
 */
bool IncoreInstrument::readChild(std::shared_ptr<ICE_IO::IHdfReadable> iHdfReadable) {
    // begin-user-code

    //If the child is null, return false
    if(iHdfReadable.get() == NULL) {
        return false;
    }

    //Cast the child into a Material - may throw a class cast exception
    try {

        //Dynamic cast to material
        Ring *newType = dynamic_cast<Ring *> (iHdfReadable.get());

        //If the type is not material, return false
        if(newType == 0) {
            return false;
        }

        //Dynamic cast up to a ring (which will preserve the clone operation on implemented dervied classes without calling new!
        std::shared_ptr<Ring> ring = std::dynamic_pointer_cast<Ring> (iHdfReadable);

        //Add ring
        IncoreInstrument::setThimble(ring);

    } catch (...) {
        return false;
    }

    //Return true for success
    return true;

    // end-user-code
}

File: src/org.eclipse.ice.reactor/src/org/eclipse/ice/reactor/pwr/IncoreInstrument.java
/*******************************************************************************
 * Copyright (c) 2012, 2014 UT-Battelle, LLC.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 *
 * Contributors:
 *   Initial API and implementation and/or initial documentation - Jay Jay Billings,
 *   Jordan H. Deyton, Dasha Gorin, Alexander J. McCaskey, Taylor Patterson,
 *   Claire Saunders, Matthew Wang, Anna Wojtowicz
 *******************************************************************************/
package org.eclipse.ice.reactor.pwr;

import java.util.ArrayList;

import org.eclipse.ice.io.hdf.IHdfReadable;
import org.eclipse.ice.io.hdf.IHdfWriteable;
import org.eclipse.ice.reactor.HDF5LWRTagType;
import org.eclipse.ice.reactor.ILWRComponentVisitor;
import org.eclipse.ice.reactor.LWRComponent;
import org.eclipse.ice.reactor.Ring;

/**
 * <p>
 * The IncoreInstrument class represents instruments (or detectors) that are
 * used for power distribution monitoring inside of a PWReactor. This class
 * contains a ring of data designed to be the "thimble" for material composition
 * on this class.
 * </p>
 * 
 * @author s4h
 */
public class IncoreInstrument extends LWRComponent {
	/**
	 * <p>
	 * An empty thimble tube used a boundary between the detector and the
	 * reactor.
	 * </p>
	 * 
	 */
	private Ring thimble;

	/**
	 * <p>
	 * The nullary Constructor.
	 * </p>
	 * 
	 */
	public IncoreInstrument() {
		// set default values
		this.setName("Instrument 1");
		this.setDescription("Default Instrument");
		this.setId(1);

		thimble = new Ring("thimble");

		this.HDF5LWRTag = HDF5LWRTagType.INCORE_INSTRUMENT;

	}

	/**
	 * <p>
	 * A parameterized Constructor.
	 * </p>
	 * 
	 * @param name
	 *            <p>
	 *            The name of this IncoreInstrument.
	 *            </p>
	 * @param thimble
	 *            <p>
	 *            An empty thimble tube used a boundary between the detector and
	 *            the reactor.
	 *            </p>
	 */
	public IncoreInstrument(String name, Ring thimble) {
		// call nullary constructor
		this();
		this.setName(name);
		this.setThimble(thimble);
	}

	/**
	 * <p>
	 * Sets an empty thimble tube used a boundary between the detector and the
	 * reactor.
	 * </p>
	 * 
	 * @param thimble
	 *            <p>
	 *            An empty thimble tube used a boundary between the detector and
	 *            the reactor.
	 *            </p>
	 */
	public void setThimble(Ring thimble) {
		// if thimble is not Null set value
		if (thimble != null) {
			this.thimble = thimble;
		}
	}

	/**
	 * <p>
	 * Returns an empty thimble tube used a boundary between the detector and
	 * the reactor.
	 * </p>
	 * 
	 * @return <p>
	 *         An empty thimble tube used a boundary between the detector and
	 *         the reactor.
	 *         </p>
	 */
	public Ring getThimble() {
		return this.thimble;
	}

	/**
	 * 
	 * @return
	 */
	public ArrayList<IHdfWriteable> getWriteableChildren() {

		// Get the children in super
		ArrayList<IHdfWriteable> children = super.getWriteableChildren();

		// If super had no children
		if (children == null) {

			// Initialize to new array list
			children = new ArrayList<IHdfWriteable>();
		}

		// Add the thimble Ring to children
		children.add(this.thimble);

		return children;
	}

	/**
	 * <p>
	 * This operation returns an ArrayList of IHdfReadable child objects. If
	 * this IHdfReadable has no IHdfReadable child objects, then null is
	 * returned.
	 * </p>
	 * 
	 * @param iHdfReadable
	 * @return
	 */
	public boolean readChild(IHdfReadable iHdfReadable) {

		// If the child is null or not an instance of LWRComponent, then return
		// false.
		if (iHdfReadable == null || !(iHdfReadable instanceof LWRComponent)) {
			return false;
		}

		// Cast the child into a LWRComponent
		LWRComponent childComponent = (LWRComponent) iHdfReadable;

		// If this is a Ring
		if (childComponent.getHDF5LWRTag() == HDF5LWRTagType.RING) {

			// Assign to correct object
			this.thimble = (Ring) childComponent;

		}

		return true;
	}

	/**
	 * <p>
	 * Overrides the equals operation to check the attributes on this object
	 * with another object of the same type. Returns true if the objects are
	 * equal. False otherwise.
	 * </p>
	 * 
	 * @param otherObject
	 *            <p>
	 *            The object to be compared.
	 *            </p>
	 * @return <p>
	 *         True if otherObject is equal. False otherwise.
	 *         </p>
	 */
	public boolean equals(Object otherObject) {
		IncoreInstrument instrument;
		boolean retVal = false;

		// If the same ref on heap, return
		if (otherObject == this) {
			return true;
		}

		// Make sure the otherObject is not null and of the same type as this
		// object
		if (otherObject != null && otherObject instanceof IncoreInstrument) {

			// Cast it
			instrument = (IncoreInstrument) otherObject;

			// check values
			retVal = (super.equals(otherObject) && this.thimble
					.equals(instrument.thimble));

		}

		// Return retVal
		return retVal;

	}

	/**
	 * <p>
	 * Returns the hashCode of the object.
	 * </p>
	 * 
	 * @return <p>
	 *         The hash of the object.
	 *         </p>
	 */
	public int hashCode() {

		// Local Declarations
		int hash = super.hashCode();

		// Local contents of the hash
		hash += 31 * this.thimble.hashCode();

		// Return the hash
		return hash;

	}

	/**
	 * <p>
	 * Deep copies the contents of the object.
	 * </p>
	 * 
	 * @param otherObject
	 *            <p>
	 *            The object to be copied.
	 *            </p>
	 */
	public void copy(IncoreInstrument otherObject) {

		// If null, return
		if (otherObject == null) {
			return;
		}

		// Copy contents - super
		super.copy(otherObject);

		// Copy local contents
		this.thimble = (Ring) otherObject.thimble.clone();

	}

	/**
	 * <p>
	 * Deep copies and returns a newly instantiated object.
	 * </p>
	 * 
	 * @return <p>
	 *         The newly instantiated copied object.
	 *         </p>
	 */
	public Object clone() {

		// Local Declarations
		IncoreInstrument instrument = new IncoreInstrument();

		// Copy contents
		instrument.copy(this);

		// Return newly instantiated object
		return instrument;

	}

	/**
	 * <p>
	 * This operation accepts an ILWRComponentVisitor that can be visit the
	 * LWRComponent to ascertain its type and perform various type-specific
	 * operations.
	 * </p>
	 * 
	 * @param visitor
	 *            <p>
	 *            The visitor
	 *            </p>
	 */
	public void accept(ILWRComponentVisitor visitor) {
		visitor.visit(this);
	}
}